airflow/jobs/scheduler_job.py

```diff
@@ -289,6 +289,45 @@ def __get_concurrency_maps(
             task_map[(dag_id, task_id)] = count
         return dag_map, task_map
 
+    @provide_session
+    def _remove_finished_tis_without_dagrun(self, session: Session = None) -> None:
+        """
+        Remove all TaskInstances for which the DagRun is missing but the task instances are in
+        finished state. Possibly the dag went missing and the task instances were failed.
+        """
+        TI = models.TaskInstance
+        query = (
+            session.query(TI)
+            .outerjoin(TI.dag_run)
+            .filter(TI.state.in_(State.finished))
+            .filter(DR.run_id.is_(None))
+        )
+
+        try:
+            count = query.count()
+            if count == 0:
+                return
+        except Exception:
+            # We want to catch any error and log it.
+            # This is to avoid the task-runner from failing.
+            self.log.error(
+                "Could not remove failed tasks for dag_id: %s",
+                ", ".join({str(ti.dag_id) for ti in query}),
+            )
+            return
+
+        query = with_row_locks(query, of=TI, session=session, **skip_locked(session=session)).all()
+        dag_ids = {ti.dag_id for ti in query}
+        if len(query) > 0:
+            session.query(TI).filter(TI.task_id.in_([ti.task_id for ti in query])).delete(
+                synchronize_session=False
+            )
```

Comment:
I’d just run `delete()`; if the query is empty, the deletion call would just be no-op anyway (and likely not much slower than the additional query implied by the `len()` call).